clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1646,9 +1646,9 @@ public void resume(Collection<TopicPartition> partitions) {
     }
 
     /**
-     * Get the last offset for the given partitions.  The last offset of a partition is the offset of the upcoming
-     * message, i.e. the offset of the last available message + 1.  If messages have never been written
-     * to the the partition, the offset returned will be 0.
+     * Get the last offset for the given partitions. The last offset of a partition is the offset of the next
```

Comment:
I am also wondering if we should mention something about the high watermark.

Reply:
Thanks for the comment, @hachikuji. What if we do something like this: 
```
Get the end offsets for the given partitions. The end offset of a partition is 
the offset of the next successfully written message. If the log is empty, 
the end offset is 0.
```
I was trying to avoid the `+1` here as I wasn't sure about the reason for it. For instance, does it matter if there is a transaction going on in this topic. If that is aborted, does the `+1` logic still apply?

Reply:
I was also wondering if this would be clearer if the high watermark was mentioned. However, it doesn't seem like we talk about the high watermark in the rest of the documentation?

Reply:
@wicknicks Hmm.. we try to cover the read_committed case below, but maybe it needs to be up front. Perhaps we could say something like this:

>  Get the end offsets for the given partitions. In the default `read_uncommitted` isolation level, the end offset is the offset of the last successfully replicated message plus one. For `read_committed` consumers, the end offset is also also limited by the lowest offset of any ongoing transaction. In other words, the end offset only advances when transactions are completed. If the log is empty, the end offset is 0.

In the rest of the Kafka documentation, we use the term "committed" to refer to messages which are have been replicated to the ISR, but we probably have to avoid that here because it can be confused with committed offsets or with the committed data of a transaction. Haha, we really need to find some synonyms.

Reply:
@hachikuji I do like moving up front the `read_committed` case. Couple of questions before I update this PR: 

* In the `read_uncommitted` case, why do we talk about what is going to happen next? This sounds more like the _next available offset_ rather than the current end offset. 
* Is the reason we want to talk about (last replicated offset + 1) instead of (next successful replication) the following: there is no guarantee that the next successfully replicated message will actually get this value? 

I am asking this since the `+1` feels a bit arbitrary. If there was a natural way to phrase it, then I would be inclined to do it that way. 

Also, for the `read_committed` case, the current documentation for this method says that the last offset is the "Last Stable Offset" LSO. It would be more precise to say this instead of saying it is limited by the lowest offset.

Reply:
> I am asking this since the +1 feels a bit arbitrary. If there was a natural way to phrase it, then I would be inclined to do it that way.

I thought it was more natural to define the log end offset in terms of the current state of the log rather than what may happen next. Technically, if we are a follower that is replicating a compacted topic, then the next message could actually be at a larger offset than the current LEO, so my phrasing is indeed more accurate. Admittedly, this technicality probably doesn't matter for the consumer.

All of this really stems from the fact that we define the log end offset to be 0 when the log is empty. From that, it follows how we must define it when it is non-empty. The choice is indeed a bit arbitrary, but I'm not sure I see how phrasing it in terms of the next message makes it less so.

> Also, for the read_committed case, the current documentation for this method says that the last offset is the "Last Stable Offset" LSO. It would be more precise to say this instead of saying it is limited by the lowest offset.

That is fine, but if we mention the LSO, then we shouldn't shy from mentioning the high watermark.